src/Pact/Types/ExpParser.hs

```diff
@@ -0,0 +1,234 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE ViewPatterns #-}
+-- |
+-- Module      :  Pact.Types.ExpParser
+-- Copyright   :  (C) 2018 Stuart Popejoy
+-- License     :  BSD-style (see the file LICENSE)
+-- Maintainer  :  Stuart Popejoy <stuart@kadena.io>
+--
+-- Parser combinator over 'Exp'.
+--
+
+module Pact.Types.ExpParser
+  where
+
+import qualified Text.Trifecta.Delta as TF
+import Control.Applicative hiding (some,many)
+import Text.Megaparsec as MP
+import Data.Proxy
+import Data.Void
+import Data.List.NonEmpty (NonEmpty(..),fromList,toList)
+import Data.List
+import Control.Monad
+import Control.Monad.State
+import Control.Arrow (second)
+import Prelude hiding (exp)
+import Data.String
+import Control.Lens hiding (prism)
+import Data.Default
+import Data.Text (Text,pack,unpack)
+import qualified Data.Text as T
+import Data.Semigroup ((<>))
+import qualified Data.Set as S
+
+import Pact.Types.Exp
+import Pact.Types.Runtime (PactError(..),PactErrorType(..))
+import Pact.Types.Info
+
+
+data Cursor = Cursor
+  { _cContext :: Maybe (Cursor,Exp Info)
+  , _cStream :: [Exp Info]
+  } deriving (Show)
+instance Default Cursor where def = Cursor def def
+
+-- | adapted from Text.Megaparsec.Stream
+
+defaultAdvance1
+  :: Pos               -- ^ Tab width (unused)
+  -> SourcePos         -- ^ Current position
+  -> t                 -- ^ Current token
+  -> SourcePos         -- ^ Incremented position
+defaultAdvance1 _width (SourcePos n l c) _t = SourcePos n l (c <> pos1)
+{-# INLINE defaultAdvance1 #-}
+
+instance Stream Cursor where
+  type Token Cursor = Exp Info
+  type Tokens Cursor = [Exp Info]
+  tokenToChunk Proxy = pure
+  tokensToChunk Proxy = id
+  chunkToTokens Proxy = id
+  chunkLength Proxy = length
+  chunkEmpty Proxy = null
+  advance1 Proxy = defaultAdvance1
+  advanceN Proxy w = foldl' (defaultAdvance1 w)
+  take1_ Cursor{..} = case _cStream of
+    [] -> Nothing
+    (t:ts) -> Just (t, Cursor _cContext ts)
+  takeN_ n s@Cursor{..}
+    | n <= 0        = Just ([], s)
+    | null _cStream = Nothing
+    | otherwise = Just $ second (Cursor _cContext) $ splitAt n _cStream
+  takeWhile_ f Cursor{..} = second (Cursor _cContext) $ span f _cStream
+
+
+data ParseState a = ParseState
+  { _psCurrent :: (Exp Info)
+  , _psUser :: a
```

Comment:
I don't get the name `psUser`. It's just any parse state the user wants to use in addition to `psCurrent`?

Also, I'm still trying to understand how it's used but I'd expect the current `Exp` to be available in a `Reader` rather than `State`.

Reply:
`enter`/`exit` manage it, I don't see how `local` could be used. The only reason it's even there is I couldn't find a way for MP to tell me what was just matched.

Reply:
Thanks @slpopejoy that answered my questions as well